flang/lib/Semantics/check-allocate.cpp

```diff
@@ -690,6 +700,20 @@ bool AllocationCheckerHelper::RunChecks(SemanticsContext &context) {
           "Object in ALLOCATE must have DEVICE attribute when STREAM option is specified"_err_en_US);
     }
   }
+  if (allocateInfo_.gotStat && allocateInfo_.statVar) {
+    if (const Symbol *symbol{allocateInfo_.statVar->symbol};
+        symbol && *ultimate_ == symbol->GetUltimate()) {
```

Comment:
This probably fails with a false positive on `allocate(a(1)%j, stat=a(2)%j)`.  Using just the last symbols won't be enough.